src/nvme/core.c

```diff
@@ -755,3 +780,95 @@ void nvme_close(struct nvme_ctrl *ctrl)
 
 	memset(ctrl, 0x0, sizeof(*ctrl));
 }
+
+int nvme_configure_cmb(struct nvme_ctrl *ctrl, bool iova)
+{
+	uint64_t cmbmsc;
+	uint32_t cmbloc;
+	uint32_t cmbsz;
+	uint64_t cap;
+
+	uint32_t lo, hi = 0;
+	uint64_t hwaddr;
+	uint64_t cba;
+	int bar;
+
+	cap = le64_to_cpu(mmio_read64(ctrl->regs + NVME_REG_CAP));
+	if (!NVME_FIELD_GET(cap, CAP_CMBS))
+		return 0;
+
+	cmbmsc = le64_to_cpu(mmio_read64(ctrl->regs + NVME_REG_CMBMSC));
+	cmbmsc |= NVME_FIELD_SET(1, CMBMSC_CRE);
+	mmio_lh_write64(ctrl->regs + NVME_REG_CMBMSC, cpu_to_le64(cmbmsc));
+
+	cmbloc = le32_to_cpu(mmio_read32(ctrl->regs + NVME_REG_CMBLOC));
+	bar = NVME_FIELD_GET(cmbloc, CMBLOC_BIR);
+	if (vfio_pci_read_config(&ctrl->pci, &lo, sizeof(lo), PCI_BASE_ADDRESS_0 + bar * 4) < 0) {
+		log_debug("could not get bar %d address\n", bar);
+		return -1;
+	}
+
+	if (lo & PCI_BASE_ADDRESS_MEM_TYPE_64) {
+		if (vfio_pci_read_config(&ctrl->pci, &hi, sizeof(hi), PCI_BASE_ADDRESS_0 + (bar + 1) * 4) < 0) {
+			log_debug("could not get bar %d address\n", bar + 1);
+			return -1;
+		}
+	}
+
+	hwaddr = (uint64_t)hi << 32 | (lo & PCI_BASE_ADDRESS_MEM_MASK);
+
+	cmbmsc = le64_to_cpu(mmio_read64(ctrl->regs + NVME_REG_CMBMSC));
+	cmbsz = le32_to_cpu(mmio_read32(ctrl->regs + NVME_REG_CMBSZ));
+
+	ctrl->cmb.bar = bar;
+	ctrl->cmb.size = nvme_cmb_size(cmbsz);
+	ctrl->cmb.hwaddr = hwaddr + NVME_FIELD_GET(cmbloc, CMBLOC_OFST);
+	ctrl->cmb.vaddr = vfio_pci_map_bar(&ctrl->pci, ctrl->cmb.bar,
+			ctrl->cmb.size, NVME_FIELD_GET(cmbloc, CMBLOC_OFST),
+			PROT_READ | PROT_WRITE);
+	if (!ctrl->cmb.vaddr) {
+		log_debug("could not map bar %d vaddr\n", bar);
+		return -1;
+	}
+
+	if (iova) {
+		if (iommu_map_vaddr(__iommu_ctx(ctrl), ctrl->cmb.vaddr,
+					ctrl->cmb.size, &ctrl->cmb.iova, 0x0)) {
+			log_debug("could not map bar vaddr to iommu\n");
+			return -1;
+		}
+
+		cba = ctrl->cmb.iova;
+		ctrl->flags |= NVME_CTRL_F_CMB_IOVA;
+	} else
+		cba = ctrl->cmb.hwaddr;
```

Comment:
I don't see any reason not to just always map the CMB into the iommu. By using the IOVA address space for the CBA, we ensure that

1. it will work if we want to do p2p
2. the cba is guaranteed not to overlap with addresses allocated in the iova space. I am not 100% sure if we are guaranteed that the hwaddr that we use for cba cannot end up being allocated by the iova allocator?